File: Hierarchical Clustering/hierarchicalClustering.py
# -*- coding: utf-8 -*-
"""
Created on Fri Jun  7 15:25:25 2019

@author: Venkatesh Ravichandran

Clustering - Hierarchical clustering

"""

#import libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

#read mall data csv
df = pd.read_csv('Mall_Customers.csv')
X = df.iloc[:,[3,4]].values

#Dendrogram for finding the optimal number of clusters
import scipy.cluster.hierarchy as sch
dendrogram = sch.dendrogram(sch.linkage(X,method='ward'))
plt.title('Dendrogram')
plt.xlabel('Customers')
plt.ylabel('Euclidean distance')
plt.show()

#from the dendrogram, we find the optimal number of clusters as 5
from sklearn.cluster import AgglomerativeClustering
clustering = AgglomerativeClustering(n_clusters = 5, affinity = 'euclidean',linkage = 'ward')
y_sch = clustering.fit_predict(X)

#Visualization of clusters
plt.scatter(X[y_sch==0,0],X[y_sch==0,1],s=100,c='red',label = 'Careful')
plt.scatter(X[y_sch==1,0],X[y_sch==1,1],s=100,c='blue',label = 'Standard')
plt.scatter(X[y_sch==2,0],X[y_sch==2,1],s=100,c='green',label = 'arget')
plt.scatter(X[y_sch==3,0],X[y_sch==3,1],s=100,c='magenta',label = 'Careless')
plt.scatter(X[y_sch==4,0],X[y_sch==4,1],s=100,c='cyan',label = 'Sensible')
plt.title('Hierarchical Clustering')
plt.xlabel('Income')
plt.ylabel('Spending Score')
plt.legend()
plt.show()

File: Hierarchical Clustering/hierarchicalClustering.R
#read mall csv data
df <- read.csv('Mall_Customers.csv')
X = df[4:5]

#Dendrogram to find the optimal number of clusters
dendrogram = hclust(dist(X,method='euclidean'),method='ward.D')
plot(dendrogram,
     main = paste('Dendrogram'),
     xlab = 'Customers',
     ylab = 'Euclidean distance')

#From dendrogram, we found the optimal number of clusters as 5

#Fit the model
hclus = hclust(dist(X,method='euclidean'),method='ward.D')
y_hc = cutree(hclus,5)

#Visualization of clusters
library(cluster)
clusplot(X, y_hc, lines = 0,shade =TRUE,
         color = TRUE, labels = 2, plotchar = FALSE,
         span = TRUE, main = paste('Cluster of clients'),
         xlab = 'Income',ylab='Purchase score')


